shorebird_code_push/lib/src/shorebird_updater.dart

```diff
@@ -0,0 +1,93 @@
+/// {@endtemplate}
+class UpdateException implements Exception {
+  /// {@macro update_exception}
+  const UpdateException(this.message);
+
+  /// The error message.
+  final String message;
+
+  @override
+  String toString() => 'UpdateException: $message';
+}
+
+/// {@template updater_state}
+/// Information about the current state of the updater.
+/// {@endtemplate}
+sealed class UpdaterState {
+  /// {@macro updater_state}
+  const UpdaterState();
+}
+
+/// {@template updater_unavailable_state}
+/// The state of the updater when the updater is not available.
+///
+/// The most common reasons for being in this state are:
+/// 1. The app is running in debug mode (Shorebird only supports release
+///    mode).
+/// 2. The app was *NOT* built using `shorebird release` and does *NOT* contain
+///    the Shorebird engine.
+/// {@endtemplate}
+class UpdaterUnavailableState extends UpdaterState {
+  /// {@macro updater_unavailable_state}
+  const UpdaterUnavailableState();
+}
+
+/// {@template updater_available_state}
+/// Information about the current state of the updater when the
+/// updater is available (e.g. the app was built with Shorebird).
+/// {@endtemplate}
+class UpdaterAvailableState extends UpdaterState {
+  /// {@macro updater_available_state}
+  const UpdaterAvailableState({
+    required this.isUpToDate,
+    required this.installedPatchNumber,
+    required this.downloadedPatchNumber,
+  });
+
+  /// Whether there are new updates available.
+  /// Returns true when there is a new patch available that has not yet been
+  /// downloaded. Otherwise, returns false.
+  final bool isUpToDate;
+
+  /// The patch number of the currently installed patch.
+  /// This is the patch that the app is currently running.
+  /// Returns null if no patch is installed.
+  final int? installedPatchNumber;
+
+  /// The patch number of the patch that has been most recently downloaded.
+  /// If no patch has been downloaded, this will be null.
+  /// See also:
+  /// * [isUpToDate] to determine whether a new update is available.
+  /// * [ShorebirdUpdater.update] to download a new patch.
+  final int? downloadedPatchNumber;
+}
+
+/// {@template shorebird_updater}
+/// Manage updates for a Shorebird app.
+/// {@endtemplate}
+abstract class ShorebirdUpdater {
+  /// {@macro shorebird_updater}
+  factory ShorebirdUpdater() => ShorebirdUpdaterImpl();
+
+  /// The current state of the updater which includes the currently installed
+  /// and downloaded patches.
+  Future<UpdaterState> get state;
+
+  /// Updates the app to the latest patch.
+  /// Note: The app must be restarted for the update to take effect.
+  ///
+  /// [onDownloadProgress] is called with the number of bytes received and the
+  /// total number of bytes to be received.
+  ///
+  /// Throws an [UpdateException] if an error occurs while updating or if no new
+  /// patch is available.
+  Future<void> update({OnDownloadProgress? onDownloadProgress});
```

Comment:
Yeah this was more of an exercise of what this could look like from the Dart side we don't have to implement it as part of this work and/or we could poll on the Dart side 🤷

Reply:
Totally fair! Just wanted to ensure you know about the limitations on the rust side.